rpcs3/Emu/Cell/lv2/sys_usbd.cpp

```diff
@@ -1,219 +1,745 @@
+	lusb_tv.tv_usec = 200;
+	
+	while (thread_ctrl::state() != thread_state::aborting && !Emu.IsStopped())
+	{
+		// Check if any device was added
+		// check_devices_vs_ldds();
+
+		// Process asynchronous requests that are pending
+		libusb_handle_events_timeout_completed(ctx, &lusb_tv, nullptr);
+		// Process fake transfers
+		{
+			auto usbh = g_idm->lock<named_thread<usb_handler_thread>>(0);
+			u64 timestamp = get_system_time() - Emu.GetPauseTime();
+
+			for (u32 tr_id = 0; tr_id < MAX_SYS_USBD_TRANSFERS; tr_id++)
+			{
+				auto& transfer = transfers[tr_id];
+				if (!transfer.busy || !transfer.fake)
+					continue;
+
+				if (transfer.expected_time > timestamp)
+					continue;
+
+				transfer.result = transfer.expected_result;
+				transfer.count = transfer.expected_count;
+				transfer.fake = false;
+				transfer.busy = false;
+
+				send_message(SYS_USBD_TRANSFER_COMPLETE, tr_id);
+			}
+		}
+		// Sleep for 1ms to avoid excessive cpu usage
+		//std::this_thread::sleep_for(1ms);
+	}
+}
+
+
+void usb_handler_thread::send_message(u32 message, u32 tr_id)
+{
+	sys_usbd.trace("Sending event: arg1=0x%x arg2=0x%x arg3=0x00", message, tr_id);
+
+	usbd_events.push({ message, tr_id, 0x00 });
+	if (receive_threads.size())
+	{
+		lv2_obj::awake(*receive_threads.front());
+		receive_threads.pop();
+	}
+}
+
+void usb_handler_thread::transfer_complete(struct libusb_transfer *transfer)
+{
+	UsbTransfer *usbd_transfer = (UsbTransfer *)transfer->user_data;
+
+	if (transfer->status != 0)
+	{
+		sys_usbd.error("Transfer Error: %d", (s32)transfer->status);
+	}
+
+	switch (transfer->status)
+	{
+	case LIBUSB_TRANSFER_COMPLETED:
+		usbd_transfer->result = HC_CC_NOERR;
+		break;
+	case LIBUSB_TRANSFER_TIMED_OUT:
+		usbd_transfer->result = EHCI_CC_XACT;
+		break;
+	case LIBUSB_TRANSFER_OVERFLOW:
+		usbd_transfer->result = EHCI_CC_BABBLE;
+		break;
+	case LIBUSB_TRANSFER_ERROR:
+	case LIBUSB_TRANSFER_CANCELLED:
+	case LIBUSB_TRANSFER_STALL:
+	case LIBUSB_TRANSFER_NO_DEVICE:
+	default:
+		usbd_transfer->result = EHCI_CC_HALTED;
+		break;
+	}
+
+	usbd_transfer->count = transfer->actual_length;
+
+	for (s32 index = 0; index < transfer->num_iso_packets; index++)
+	{
+		u8 iso_status;
+		switch (transfer->iso_packet_desc[index].status)
+		{
+		case LIBUSB_TRANSFER_COMPLETED:
+			iso_status = USBD_HC_CC_NOERR;
+			break;
+		case LIBUSB_TRANSFER_TIMED_OUT:
+			iso_status = USBD_HC_CC_XACT;
+			break;
+		case LIBUSB_TRANSFER_OVERFLOW:
+			iso_status = USBD_HC_CC_BABBLE;
+			break;
+		case LIBUSB_TRANSFER_ERROR:
+		case LIBUSB_TRANSFER_CANCELLED:
+		case LIBUSB_TRANSFER_STALL:
+		case LIBUSB_TRANSFER_NO_DEVICE:
+		default:
+			iso_status = USBD_HC_CC_MISSMF;
+			break;
+		}
+
+		usbd_transfer->iso_request.packets[index] = ((iso_status & 0xF) << 12 | (transfer->iso_packet_desc[index].actual_length & 0xFFF));
+	}
+
+	usbd_transfer->busy = false;
+
+	send_message(SYS_USBD_TRANSFER_COMPLETE, usbd_transfer->transfer_id);
+
+	if (sys_usbd.enabled == logs::level::trace)
```

Comment:
If you pass this string constructor as an immediatly executed lambda to the logging you won't need to check log level manually.

Reply:
Nice, I didn't realize that, I will change that :).